crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3564,6 +3564,63 @@ impl<'db> Type<'db> {
             _ => KnownClass::Str.to_instance(db),
         }
     }
+
+    /// Returns where this type is defined.
+    ///
+    /// It's the foundation for the editor's "Go to type definition" feature
+    /// where the user clicks on a value and it takes them to where the value's type is defined.
+    ///
+    /// This method returns `None` for unions and intersections because how these
+    /// should be handled, especially when some variants don't have definitions, is
+    /// specific to the call site.
+    pub fn definition(&self, db: &'db dyn Db) -> Option<TypeDefinition<'db>> {
+        match self {
+            Type::BoundMethod(method) => {
+                Some(TypeDefinition::Function(method.function(db).definition(db)))
+            }
+            Type::FunctionLiteral(function) => {
+                Some(TypeDefinition::Function(function.definition(db)))
+            }
+            Type::ModuleLiteral(module) => Some(TypeDefinition::Module(module.module(db))),
+            Type::ClassLiteral(class_literal) => {
+                Some(TypeDefinition::Class(class_literal.class().definition(db)))
+            }
+            Type::Instance(instance) => {
+                Some(TypeDefinition::Class(instance.class().definition(db)))
+            }
+            Type::KnownInstance(instance) => match instance {
+                KnownInstanceType::TypeVar(var) => {
```

Comment:
Orthogonal to this PR but I just noticed in playground that we don't implement goto-type-definition for PEP 695 type aliases (`type Foo = int`), even though we could jump to the `type` statement. I think this would require keeping a reference to the `Definition` in `TypeAliasType`.